src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1022,250 +1015,222 @@ private ImmutableArray<BoundNode> ConvertCollectionExpressionElements(
+                BoundObjectOrCollectionValuePlaceholder? implicitReceiver,
+                CollectionInitializerAddMethodBinder? collectionInitializerAddMethodBinder,
+                BindingDiagnosticBag diagnostics)
+            {
+                if (elementKeyValueTypes is (var elementKeyType, var elementValueType) &&
+                    elementConversion.TryGetKeyValueConversions(out var keyConversion, out var valueConversion))
                 {
-                    BoundNode convertedElement;
-                    switch (element)
+                    if (!useIndexer)
                     {
-                        case BoundCollectionExpressionWithElement withElement:
-                            continue;
-                        case BoundCollectionExpressionSpreadElement spreadElement:
-                            convertedElement = bindSpreadElement(
-                                spreadElement,
-                                elementType,
-                                elementConversions[conversionIndex++],
-                                diagnostics);
-                            break;
-                        case BoundExpression expressionElement:
-                            convertedElement = CreateConversion(
-                                element.Syntax,
-                                expressionElement,
-                                elementConversions[conversionIndex++],
-                                isCast: false,
-                                conversionGroupOpt: null,
-                                destination: elementType,
-                                diagnostics);
-                            break;
-                        default:
-                            throw ExceptionUtilities.UnexpectedValue(element);
+                        _ = binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__ctor, diagnostics, syntax: expressionSyntax);
+                    }
+                    _ = binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__get_Key, diagnostics, syntax: expressionSyntax);
+                    _ = binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__get_Value, diagnostics, syntax: expressionSyntax);
+                    if (!ConversionsBase.IsKeyValuePairType(binder.Compilation, expressionElement.Type, out var keyType, out var valueType))
+                    {
+                        throw ExceptionUtilities.UnexpectedValue(expressionElement.Type);
+                    }
+                    BoundExpression convertedExpression = binder.BindToNaturalType(expressionElement, diagnostics);
+                    var keyPlaceholder = new BoundValuePlaceholder(expressionSyntax, keyType);
+                    var valuePlaceholder = new BoundValuePlaceholder(expressionSyntax, valueType);
+                    var keyValuePairConversion = new BoundKeyValuePairConversion(
```

Comment:
It looks like we create this `BoundKeyValuePairConversion` even when we only have identity conversions. Then it seems `VisitKeyValuePairConversion` would produce lowered code that extracts the key and value parts, then reconstructs an identical KVP from it. Is that the case?

Reply:
If there is an implicit conversion from the source `KeyValuePair<K1, V1>` to the destination `KeyValuePair<K2, V2>`, we'll use that implicit conversion and convert between `KeyValuePair<,>` instances directly rather than deconstructing, converting, and reconstructing parts. See `Conversions.GetCollectionExpressionElementConversion()` which checks for implicit conversion between `KeyValuePair<,>` types first before checking parts, and only generates a `ConversionKind.KeyValuePair` for the latter case.